### PR TITLE
Task 3. Routing configuration Task 4. Custom Middleware Task 5. Filters

### PR DESCRIPTION
**Pull request by CerenSusuz**
## [Task 3. Routing configuration Task 4. Custom Middleware Task 5. Filters](https://learn.epam.com/myLearning/program?groupGuid=44c7a385-8be5-48e2-b659-c458e8b6af60&tab=panels) 

Task 3. Routing configuration

Configure routing to enable getting images by not only the standard path {controller}/{action}/…, but also by images/{image_id}

Task 4. Custom Middleware

Create your own middleware component for image caching. Middleware should:

Check Content-Type for every response, and, if any image of a valid format returned:
Keep the image on the disk (as file)
If subsequent requests access the same image, return it from the cache directory
Support the following options:
The path for the cache directory
Max count of cached images 
Cache expiration time (if no requests during this time, cache cleaned) 
Include the middleware into the request pipeline of your application

Task 5. Filters

Add MVC filter for logging calls to Actions. The filter should:

Log Action start/end 
Use standard ASP.Net Core logging infrastructure 
Provide an option to turn on/off logging parameters of the invoked Action method (off by default)

* Create a new CustomMiddleware on Middleware folder
* Create a new CustomFilter on Filters folder
* Create a new CustomExtensions on Extensions folder
* Create a new PAge for showing just the selected image

* Tested locally.

* on categories page you can access all images of categories with the Show Images button

image

* on Images page you can see all image with their category name

image

* if you just want to see selected image, click show image button and the image will come

image

* log action filter result on output

image

### CODE REVIEW

**Review by dotnet-fizzyy — CHANGES_REQUESTED** [review]
Unfortunately, your PR does not meet ACs of tasks, please, read them carefully

File: GetandTake/Business/Services/Concrete/SupplierManager.cs [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
 public class SupplierManager : ISupplierService
 {
     private readonly ISupplierRepository _repository;
 
-    public SupplierManager(ISupplierRepository repository) => 
+    public SupplierManager(ISupplierRepository repository) =>
         _repository = repository;
 
-    public async Task CreateAsync(Supplier supplier) => 
+    [CacheRemoveAspect("ISupplierService.Get")]
```
Comment (dotnet-fizzyy) [inline GetandTake/Business/Services/Concrete/SupplierManager.cs]: You may easily use `nameof` section to get method name:
```suggestion
    [CacheRemoveAspect(nameof(ISupplierService.Get))]
```

Please, do it for the rest of methods

File: GetandTake/Core/Aspects/Caching/CacheAspect.cs [hunk 2]
```diff
@@ -0,0 +1,42 @@
+﻿using Castle.DynamicProxy;
+using GetandTake.Core.CrossCuttingConcerns.Caching;
+using GetandTake.Core.Utilities.Interceptors;
+using GetandTake.Core.Utilities.IoC;
+
+namespace GetandTake.Core.Aspects.Caching;
+
+public class CacheAspect : MethodInterception
+{
+    private const int _defaultDuration = 60;
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Aspects/Caching/CacheAspect.cs]: Please, always use _PascalCase_ for constants despite their modifier

File: GetandTake/Core/Filters/LogActionFilter.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
+﻿using Microsoft.AspNetCore.Mvc.Filters;
+using System.Diagnostics;
+
+namespace GetandTake.Core.Filters;
+
+public class LogActionFilter : ActionFilterAttribute
+
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Filters/LogActionFilter.cs]: This space is not needed

File: GetandTake/Core/Filters/LogActionFilter.cs [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+{
+    public override void OnActionExecuting(ActionExecutingContext filterContext)
+    {
+        Log("OnActionExecuting", filterContext.RouteData);
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Filters/LogActionFilter.cs]: Let's use `nameof` to get method name instead of hardcoding it

File: GetandTake/Core/Filters/LogActionFilter.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    public override void OnResultExecuted(ResultExecutedContext filterContext)
+    {
+        Log("OnResultExecuted", filterContext.RouteData);
+    }
+
+    private void Log(string methodName, RouteData routeData)
+    {
+        var pageName = routeData.Values["page"];
+        var message = string.Format("{0} page:{1}", methodName, pageName);
+        Debug.WriteLine(message, "Action Filter Log");
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Filters/LogActionFilter.cs]: Let's inject `ILogger` and use it instead of using diagnostic tools from C#.
On practice, there are many cases when incoming requests are writing to files/cloud streams and custom logging configuration allows that.

File: GetandTake/Core/Utilities/Interceptors/AspectInterceptorSelector.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public IInterceptor[] SelectInterceptors(Type type, MethodInfo method, IInterceptor[] interceptors)
+    {
+        var classAttributes = type.GetCustomAttributes<MethodInterceptionBaseAttribute>
+            (true).ToList();
+        var methodAttributes = type.GetMethod(method.Name)
+            .GetCustomAttributes<MethodInterceptionBaseAttribute>(true);
+        classAttributes.AddRange(methodAttributes);
+
+        return classAttributes.OrderBy(x => x.Priority).ToArray();
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Utilities/Interceptors/AspectInterceptorSelector.cs]: Let's give more descriptive name for lambda argument

File: GetandTake/Core/Utilities/Interceptors/MethodInterception.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace GetandTake.Core.Utilities.Interceptors;
+
+public abstract class MethodInterception : MethodInterceptionBaseAttribute
+{
+    protected virtual void OnBefore(IInvocation invocation) { }
+    protected virtual void OnAfter(IInvocation invocation) { }
+    protected virtual void OnException(IInvocation invocation, Exception e) { }
+    protected virtual void OnSuccess(IInvocation invocation) { }
+    public override void Intercept(IInvocation invocation)
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Utilities/Interceptors/MethodInterception.cs]: Please, separate your implementations with empty lines to avoid items overlapping each other

File: GetandTake/Program.cs [hunk 8]
```diff
@@ -29,36 +42,33 @@
...
 
-
 var app = builder.Build();
 // Configure the HTTP request pipeline.
 if (!app.Environment.IsDevelopment())
 {
     app.UseExceptionHandler("/Error");
     // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
     app.UseHsts();
+    app.ExceptionHandler();
```
Comment (dotnet-fizzyy) [inline GetandTake/Program.cs]: It is recommended to place exception handler in general middleware pipeline, let's revert this change

File: GetandTake/Pages/Categories/Images.cshtml.cs [hunk 9]
```diff
@@ -1,23 +1,23 @@
...
-    public class ImagesModel : PageModel
-    {
-        private readonly ICategoryService _categoryService;
+    private readonly ICategoryService _categoryService;
+
+    public IEnumerable<Category> Categories { get; private set; }
 
-        public IEnumerable<Category> Categories { get; private set; }
+    public ImagesModel(
+        ICategoryService categoryService)
```
Comment (dotnet-fizzyy) [inline GetandTake/Pages/Categories/Images.cshtml.cs]: No need to wrap single line on the next row

File: GetandTake/Pages/Categories/Image.cshtml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+﻿@page
+@model GetandTake.Pages.Categories.ImageModel
+@if (Model.Category.ImagePath != null)
+{
+    string imagePath = $"https://localhost:7005/" + Model.Category.ImagePath;
+    <img src="@imagePath" class="img-thumbnail" width="250px;" />
+}
+
+
```
Comment (dotnet-fizzyy) [inline GetandTake/Pages/Categories/Image.cshtml]: Guess these spaces are not required

File: GetandTake/Pages/Categories/Image.cshtml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+﻿@page
+@model GetandTake.Pages.Categories.ImageModel
+@if (Model.Category.ImagePath != null)
```
Comment (dotnet-fizzyy) [inline GetandTake/Pages/Categories/Image.cshtml:4]: Let's add empty line before `@if`, please, to prevent code overlapping

File: GetandTake/Pages/Categories/Image.cshtml [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+{
+    string imagePath = $"https://localhost:7005/" + Model.Category.ImagePath;
```
Comment (dotnet-fizzyy) [inline GetandTake/Pages/Categories/Image.cshtml]: Not a good idea to hardcode your URL here: if you decide to change host, it will cause issues.
You can specify URL in app settings and take it from there

File: GetandTake/Middlewares/CachingMiddleware.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+{
+    private readonly RequestDelegate _next;
+    private readonly ILogger _logger;
+
+    public CachingMiddleware(
+        RequestDelegate next,
+        ILoggerFactory logFactory)
+    {
+        _next = next;
+        _logger = logFactory.CreateLogger("CachingMiddleware");
```
Comment (dotnet-fizzyy) [inline GetandTake/Middlewares/CachingMiddleware.cs]: `nameof`

File: GetandTake/Core/Utilities/Interceptors/AspectInterceptorSelector.cs [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+using System.Reflection;
+
+namespace GetandTake.Core.Utilities.Interceptors;
+
+public class AspectInterceptorSelector : IInterceptorSelector
+{
+    public IInterceptor[] SelectInterceptors(Type type, MethodInfo method, IInterceptor[] interceptors)
+    {
+        var classAttributes = type.GetCustomAttributes<MethodInterceptionBaseAttribute>
+            (true).ToList();
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Utilities/Interceptors/AspectInterceptorSelector.cs]: Let's use named parameters instead of hardcoding values. For some devs, not familiar with code base or 3rd party solutions it will be difficult to understand the purpose of this value

**Review by dotnet-fizzyy — COMMENTED** [review]
Please, make sure your filter satisfies the following condition:

* Provide an option to turn on/off logging parameters of the invoked Action method (off by default)

**Comment by CerenSusuz** [comment]
> I couldn't understand that one.

Provide an option to turn on/off logging parameters of the invoked Action method (off by default)

**Comment by dotnet-fizzyy** [comment]
> Provide an option to turn on/off logging parameters of the invoked Action method (off by default)

You should pass from "app settings.json" variable that is responsible for logging in filter, it means, filter should get this variable and check its status: if it is true - execute logging, otherwise skip it :)

File: GetandTake/Core/Aspects/Caching/CacheAspect.cs [hunk 15]
```diff
@@ -0,0 +1,42 @@
+﻿using Castle.DynamicProxy;
+using GetandTake.Core.CrossCuttingConcerns.Caching;
+using GetandTake.Core.Utilities.Interceptors;
+using GetandTake.Core.Utilities.IoC;
+
+namespace GetandTake.Core.Aspects.Caching;
+
+public class CacheAspect : MethodInterception
+{
+    private const int defaultDuration = 60;
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Aspects/Caching/CacheAspect.cs]: Please, use _PascalCase_ for constants, it should be:
```suggestion
    private const int DefaultDuration = 60;
```

File: GetandTake/Core/Filters/LogActionFilterAttribute.cs [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+    private readonly ILogger<LogActionFilterAttribute> _logger;
+
+    public LogActionFilterAttribute(ILogger<LogActionFilterAttribute> logger)
+    {
+        _logger = logger;
+    }
+
+    public override void OnActionExecuting(ActionExecutingContext filterContext)
+    {
+        _logger.LogInformation(nameof(OnActionExecuting), filterContext.RouteData.Values["page"]);
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Filters/LogActionFilterAttribute.cs]: Since `filterContext.RouteData.Values["page"]` repeats several times, it makes sense to move the logic of getting this variable into function. Also, each method in filter has its own class for `filterContext`. In this way, you can simply pass `RouteData` class into your function and get `page` value from here. Hope you got my point ;)

File: GetandTake/Core/Utilities/Interceptors/AspectInterceptorSelector.cs [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+namespace GetandTake.Core.Utilities.Interceptors;
+
+public class AspectInterceptorSelector : IInterceptorSelector
+{
+    public IInterceptor[] SelectInterceptors(Type type, MethodInfo method, IInterceptor[] interceptors)
+    {
+        var classAttributes = type.GetCustomAttributes<MethodInterceptionBaseAttribute>
+            (inherit: true).ToList();
+        var methodAttributes = type.GetMethod(method.Name)
+            .GetCustomAttributes<MethodInterceptionBaseAttribute>(true);
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Utilities/Interceptors/AspectInterceptorSelector.cs]: Same case with named param here as well, please

File: GetandTake/appsettings.json [hunk 18]
```diff
@@ -6,10 +6,16 @@
...
     "DefaultConnection": "Server=EPTRANKW0038\\SQLEXPRESS;Database=Northwind;Trusted_Connection=True;"
   },
-  "products": {
+  "Products": {
     "maximumAmount": 0
+  },
+  "Host": {
+    "HostUrl": "https://localhost:7005/"
+  },
+  "LogActionFilter": {
```
Comment (dotnet-fizzyy) [inline GetandTake/appsettings.json]: To make it more abstract and reusable in future let's make name of this section like `Logging` or `LoggingParameters`

File: GetandTake/Program.cs [hunk 19]
```diff
@@ -1,14 +1,28 @@
...
 
 var builder = WebApplication.CreateBuilder(args);
 builder.RegisterLogging();
 // Add services to the container.
 builder.Services
     .AddRazorPages()
+    //.AddMvcOptions(options =>
+    //{
+    //    options.Filters.Add(typeof(LogActionFilterAttribute));
+    //})
```
Comment (dotnet-fizzyy) [inline GetandTake/Program.cs]: Is it necessary to comment this section?

File: GetandTake/Core/Filters/LogActionFilterAttribute.cs [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+        _appSettings = appSettings;
+    }
+
+    public async Task OnPageHandlerExecutionAsync(
+        PageHandlerExecutingContext context,
+        PageHandlerExecutionDelegate next)
+    {
+        var page = context.RouteData.Values["page"];
+
+        if (_appSettings.LogFilter.IsLogFilterActive == true)
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Filters/LogActionFilterAttribute.cs]: You can just simplify it:
```suggestion
        if (_appSettings.LogFilter.IsLogFilterActive)
```

File: GetandTake/Core/Filters/LogActionFilterAttribute.cs [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+        {
+            _logger.LogInformation(nameof(context), page);
+            await next.Invoke();
```
Comment (dotnet-fizzyy) [inline GetandTake/Core/Filters/LogActionFilterAttribute.cs]: Well, I am afraid if your condition will be `false` then this middleware chain method will not be executed. It might seriously affect your app workflow. Try to always put `await next.Invoke();` after some logic filter methods (at the end of the method I wanted to say)

File: GetandTake/Program.cs (hunk 19 above)
Comment (CerenSusuz) [reply GetandTake/Program.cs]: I forgot to un-comment that part. :)

File: GetandTake/Core/Filters/LogActionFilterAttribute.cs (hunk 21 above)
Comment (CerenSusuz) [reply GetandTake/Core/Filters/LogActionFilterAttribute.cs]: In this case: you mean that should be like that:
       
        if (_appSettings.LogFilter.IsLogFilterActive)
        {
            _logger.LogInformation(nameof(context), page);
        }
        await next.Invoke();

Comment (dotnet-fizzyy) [reply GetandTake/Core/Filters/LogActionFilterAttribute.cs]: Yes, this is correct :)

**Review by dotnet-fizzyy — APPROVED** [review]
